src/QirRuntime/public/ResourceStatistics.hpp

```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include <sstream>
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    static const char* names[] = {"cX", "cCX", "cY",  "cCY",  "cZ", "cCZ", "cH",  "cCH",  "cR", "cCR",
+                                  "cS", "cCS", "cAS", "cCAS", "cT", "cCT", "cAT", "cCAT", "cM"};
+    static constexpr int counters = sizeof(names) / sizeof(char*);
+
+    struct ResourceStatistics
+    {
+        // these members must match names[] above
+        long cX, cCX = 0;
+        long cY, cCY = 0;
+        long cZ, cCZ = 0;
+        long cH, cCH = 0;
+        long cR, cCR = 0;
+        long cS, cCS, cAS, cCAS = 0;
+        long cT, cCT, cAT, cCAT = 0;
+        long cM = 0;
+        //=========================================
+
+        long cQubitWidth = 0;
+        long cQubits = 0;
+
+        long cCircuitDepth = 0;
+
+        // {"qubit_width":3, "circuit_depth": 2, "statistics":[{"metric":"cX","value":1},{"metric":"cR","value":2}]}
+        // qubit_width is required even if it's equal zero, other metrics reported only if non-zero.
+        void Print(std::ostringstream& os)
+        {
+            os << "{";
+            os << "\"qubit_width\":" << cQubitWidth;
+
+            if (cCircuitDepth > 0)
+            {
+                os << ",\"circuit_depth\":" << cCircuitDepth;
+            }
+
+            os << ",\"statistics\":[";
+            long* fields = reinterpret_cast<long*>(this);

```

Comment:
This is massively unsafe and error-prone, especially if I'm brave enough to derive from this class. There's no guarantee that there's nothing between what `this` points to and the first instance variable, and there's no guarantee that the instance variables wind up in memory in the same order they appear in the type. While one compiler may work that way, another might not, or the same compiler might work differently on different platforms or with different optimization options.
Either you should step through the fields explicitly, without a loop, in which case you don't need the string array and can avoid that fragility, or replace the whole thing with a `map` that you can iterate over cleanly.

Reply:
I cannot use a map because it will surely prevent LLVM from optimizing increments of the counters. I could replace the field names with an enum and an array of longs which would make it somewhat safer, but I don't know whether that would play well with LLVM's optimization passes.